rustdoc: use focus for search navigation

jsha opened the pull request:
Rather than keeping track of highlighted element inside the JS, take advantage of `.focus()` and the :focus CSS pseudo-class.

This required wrapping each row of results in one big `<a>` tag (because anchors can be focused, but table rows cannot). That in turn required moving from a table layout to a div layout with float.

This makes it so Ctrl+Enter opens links in new tabs, and using the arrow keys to navigate off the bottom of the page scrolls the rest of the page into view. It also simplifies the keyboard event handling. It eliminates the need for click handlers on the search results, and for tracking mouse movements.

This changes the UI treatment of mouse hovering. A hovered element now gets a light grey background, but does not change the focus. It's possible to have two highlighted search results: one that is focused (via keyboard) and one that is hovered (via mouse). Pressing enter will activate the focused link; clicking will activate the hovered link. This matches up with how Firefox and Chrome handle suggestions in their URL bar, and avoids stray mouse movements changing the focus.

Selecting tabs is now done with left/right arrows while any search result is focused. The visibility of results on each search tab is
controlled with the "active" class, rather than by setting display: none directly. Note that the old code kept track of highlighted search element when tabbing back and forth. The new code doesn't.

Demo at https://hoffman-andrews.com/rust/focus-search-results2/std/?search=fn

Fixes #84384
Fixes #79962
Fixes #79872

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

Some changes occurred in HTML/CSS themes.

cc @GuillaumeGomez

rust-highfive commented:
r? @ollie27

(rust-highfive has picked a reviewer for you, use r? to override)

jsha commented:
@bors r? @GuillaumeGomez 

BTW I see in https://github.com/rust-lang/rust/issues/65212#issuecomment-741952935 that a desired functionality is that you can press down arrow a few times, and then type a letter key to add to your search. That doesn't work in the current revision but should be easy to restore: We can make it so that typing any letter key while the search results are highlight focuses the search input and inserts that letter.

jsha commented:
Also cc @cratelyn and @Manishearth who had great feedback about keyboard navigation in #65212.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2415544296) [(plain)](https://github.com/rust-lang/rust/commit/08543cb11780bc7c863bcef73b6cf8944fde86a4/checks/2415544296/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11770
.................................................................................................... 9500/11770
...................................................................................................i 9600/11770
......i............................................................................................. 9700/11770
..............................................iiiiiii.iiiiii.i...................................... 9800/11770
.................................................................................................... 10000/11770
.................................................................................................... 10100/11770
.................................................................................................... 10200/11770
.................................................................................................... 10300/11770
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 35 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii....
test result: ok. 4 passed; 0 failed; 31 ignored; 0 measured; 0 filtered out; finished in 0.12s

 finished in 0.184 seconds
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.55s

 finished in 2.622 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
   Compiling rustdoc-themes v0.1.0 (/checkout/src/tools/rustdoc-themes)
    Finished release [optimized] target(s) in 0.68s
rustdoc: [check-theme] Starting tests! (Ignoring all other arguments)
 - Checking "/checkout/src/librustdoc/html/static/themes/dark.css"... FAILED
  Missing ".content a.result-fn:focus,.content a.result-method:focus,.content a.result-tymethod:focus" rule
  Missing ".content a.result-type:focus" rule
  Missing ".content a.result-keyword:focus" rule
  Missing ".content a.result-constant:focus,.content a.result-static:focus" rule
  Missing ".content a.result-mod:focus,.content a.result-externcrate:focus" rule
  Missing ".content a.result-attr:focus,.content a.result-derive:focus,.content a.result-macro:focus" rule
  Missing ".content a.result-primitive:focus" rule
  Missing ".content a.result-struct:focus" rule
  Missing ".content a.result-union:focus" rule
  Missing ".content a:focus" rule
  Missing ".content a.result-traitalias:focus" rule
  Missing ".content a.result-foreigntype:focus" rule
  Missing ".content a.result-trait:focus" rule
  Missing ".content a.result-enum:focus" rule
  Missing ".content a:focus,.content span:focus" rule
  Missing ".content a:hover" rule
 - Checking "/checkout/src/librustdoc/html/static/themes/ayu.css"... FAILED
  Missing ".content a.result-fn:focus,.content a.result-method:focus,.content a.result-tymethod:focus" rule
  Missing ".content a.result-type:focus" rule
  Missing ".content a.result-keyword:focus" rule
  Missing ".content a.result-constant:focus,.content a.result-static:focus" rule
  Missing ".content a.result-mod:focus,.content a.result-externcrate:focus" rule
  Missing ".content a.result-attr:focus,.content a.result-derive:focus,.content a.result-macro:focus" rule
  Missing ".content a.result-primitive:focus" rule
  Missing ".content a.result-struct:focus" rule
  Missing ".content a.result-union:focus" rule
  Missing ".content a:focus" rule
  Missing ".content a.result-traitalias:focus" rule
  Missing ".content a.result-foreigntype:focus" rule
  Missing ".content a.result-trait:focus" rule
  Missing ".content a.result-enum:focus" rule
  Missing ".content a:focus,.content span:focus" rule
  Missing ".content a:hover" rule

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rustdoc-themes" "/checkout/obj/build/bootstrap/debug/rustdoc" "/checkout/src/librustdoc/html/static/themes"
expected success, got: exit code: 1

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2417132057) [(plain)](https://github.com/rust-lang/rust/commit/b44c146a9b1d347f78ad4c26c6e40c0e99623af8/checks/2417132057/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11770
.................................................................................................... 9500/11770
...................................................................................................i 9600/11770
......i............................................................................................. 9700/11770
.............................................iiiiiii..iiiiii.i...................................... 9800/11770
.................................................................................................... 10000/11770
.................................................................................................... 10100/11770
.................................................................................................... 10200/11770
.................................................................................................... 10300/11770
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 35 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii....

 finished in 0.179 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i..i..ii....i.i....ii..........ii.ii........i.....i....i......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.45s

 finished in 2.518 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
   Compiling rustdoc-themes v0.1.0 (/checkout/src/tools/rustdoc-themes)
    Finished release [optimized] target(s) in 0.59s
rustdoc: [check-theme] Starting tests! (Ignoring all other arguments)
 - Checking "/checkout/src/librustdoc/html/static/themes/dark.css"... FAILED
  Missing ".content a.result-fn:focus,.content a.result-method:focus,.content a.result-tymethod:focus" rule
  Missing ".content a.result-type:focus" rule
  Missing ".content a.result-keyword:focus" rule
  Missing ".content a.result-constant:focus,.content a.result-static:focus" rule
  Missing ".content a.result-mod:focus,.content a.result-externcrate:focus" rule
  Missing ".content a.result-attr:focus,.content a.result-derive:focus,.content a.result-macro:focus" rule
  Missing ".content a.result-primitive:focus" rule
  Missing ".content a.result-struct:focus" rule
  Missing ".content a.result-union:focus" rule
  Missing ".content a:focus" rule
  Missing ".content a.result-traitalias:focus" rule
  Missing ".content a.result-foreigntype:focus" rule
  Missing ".content a.result-trait:focus" rule
  Missing ".content a.result-enum:focus" rule
  Missing ".content a:focus,.content span:focus" rule
  Missing ".content a:hover" rule
 - Checking "/checkout/src/librustdoc/html/static/themes/ayu.css"... FAILED
  Missing ".content a.result-fn:focus,.content a.result-method:focus,.content a.result-tymethod:focus" rule
  Missing ".content a.result-type:focus" rule
  Missing ".content a.result-keyword:focus" rule
  Missing ".content a.result-constant:focus,.content a.result-static:focus" rule
  Missing ".content a.result-mod:focus,.content a.result-externcrate:focus" rule
  Missing ".content a.result-attr:focus,.content a.result-derive:focus,.content a.result-macro:focus" rule
  Missing ".content a.result-primitive:focus" rule
  Missing ".content a.result-struct:focus" rule
  Missing ".content a.result-union:focus" rule
  Missing ".content a:focus" rule
  Missing ".content a.result-traitalias:focus" rule
  Missing ".content a.result-foreigntype:focus" rule
  Missing ".content a.result-trait:focus" rule
  Missing ".content a.result-enum:focus" rule
  Missing ".content a:focus,.content span:focus" rule
  Missing ".content a:hover" rule

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rustdoc-themes" "/checkout/obj/build/bootstrap/debug/rustdoc" "/checkout/src/librustdoc/html/static/themes"
expected success, got: exit code: 1

```

</details>

GuillaumeGomez commented:
That's what I was a bit afraid of: scroll with arrows isn't working anymore, we have to use the mouse now. I'm a bit shared on this. Also: can you "restore" the previous coloring for the focused search result please?

Another important point: when we go back to the search input, can you scroll to the top too and not simply focus on the input? We add some issues about this behavior because some browsers only care if the element is in the viewport and not if it's reachable/visible (because hidden under other elements). That could be an issue on docs.rs for example.

> BTW I see in #65212 (comment) that a desired functionality is that you can press down arrow a few times, and then type a letter key to add to your search. That doesn't work in the current revision but should be easy to restore: We can make it so that typing any letter key while the search results are highlight focuses the search input and inserts that letter.

That'd be nice to keep this behaviour, indeed. Question though: what happens if we press the letter 's'? Should it only focus the search input (like it does when it's not focused) or should it also enter the letter 's'?

> Selecting tabs is now done with left/right arrows while any search result is focused. The visibility of results on each search tab is controlled with the "active" class, rather than by setting display: none directly. Note that the old code kept track of highlighted search element when tabbing back and forth. The new code doesn't.

So three things here:
 1. Being able to switch tabs when the search input is focused (using the current shortcut CTRL+up/down) would be nice, because for now it'd force to have a search result focused, which isn't super intuitive.
 2. If so, having two different shortcuts for a same behaviour seems really bad, no idea how to make it work otherwise though...
 3. Please keep track of the focused element in the other tabs.

One last point (small bug apparently): this search triggers a js error: https://hoffman-andrews.com/rust/focus-search-results2/std/?search=-%3E%20str . I was trying to check how it handled things in case the current tab was empty. The current behaviour is that it selects the first non-empty.

Also, I couldn't check but: does it keep "memory" of the last selected tab when you do another search (without selecting a result I mean, just doing another search when on search results).

Manishearth commented:
> That's what I was a bit afraid of: scroll with arrows isn't working anymore, we have to use the mouse now. I'm a bit shared on this. Also: can you "restore" the previous coloring for the focused search result please?

Scroll with arrows works fine for me in this prototype? It's actually _better_ now since it's per-row instead of per-cell.

> Being able to switch tabs when the search input is focused (using the current shortcut CTRL+up/down) would be nice, because for now it'd force to have a search result focused, which isn't super intuitive.

Idk, for me it's quite intuitive, you hit the down arrow once to enter the search results, and then you can use all four arrow keys to naturally navigate. This is how most tab-based keyboard navigation works.


> Please keep track of the focused element in the other tabs.

I'm against this: this is overriding browser default focus behavior with JS and I have never seen that go well. Fine to override focus to the start of the list, but custom stuff tends to break down, and is usually bad for accessibility.

GuillaumeGomez commented:
> Oh I see what you mean. That's a fair point, I guess users can use page up/page down, maybe.

I'm sad to say this but: those keys aren't available on all computers... XD (not even joking here)

Manishearth commented:
> I'm sad to say this but: those keys aren't available on all computers... XD (not even joking here)

I'm aware, usually it exists behind a Fn key (Fn-arrow gives home/end/pgup/pgdn), but yeah, this isn't ideal. Not sure what the best option is here.

> I'd argue that this is not the default behavior considering that we're creating it. And also that this changes the current behaviour. Not a blocker but something we should be careful about.


Browser focus navigation continues to work, which is the default behavior people will be used to more than site-specific keyboard shortcuts.

> It's very common for me to switch between two tabs and it'd be annoying to have to go through all the items again. Not sure what the accessibility issue here would be though. More explanations please? :)

In general trying to replace existing browser features with custom JS never goes well in my experience, and ends up with things breaking. A lot of modern websites do this kind of thing a lot and it breaks down pretty quickly and is hard to keep cross-browser compatible.

GuillaumeGomez commented:
> In general trying to replace existing browser features with custom JS never goes well in my experience, and ends up with things breaking. A lot of modern websites do this kind of thing a lot and it breaks down pretty quickly and is hard to keep cross-browser compatible.

The only thing the JS would do would be to re-focus the element that was previously focused in the tab when we switch between them. We are not overriding more than what we are already. We already use JS to switch tabs, so what would that change to re-focus the previously focused element in that tab?

Manishearth commented:
> The only thing the JS would do would be to re-focus the element that was previously focused in the tab when we switch between them. We are not overriding more than what we are already. 

Because we're trying to keep track of focus and remember it as well, as opposed to just _setting_ it. I guess it's fine but in my experience this kind of thing doesn't go well.

I'd rather defer to @jsha on this though.

Manishearth commented:
> What I see is this: search input or search results focused -> down arrow selects a result. No focus -> down arrow scrolls the page.
> 
> >

This is what I see too. I _think_ what @GuillaumeGomez is referring to is that previously you could scroll the entire search results page with the arrow key. I'm ... okay with losing that functionality when there's search focus.

jsha commented:
> previously you could scroll the entire search results page with the arrow key. 

What I see on https://doc.rust-lang.org/std/?search=fn is:

 - When search is focused, you _can't_ scroll the page with the arrow keys because arrow keys highlight search results.
 - When nothing is focused (i.e. click in a blank part of the page), you can scroll the page with the arrow keys.
 
 But I can reproduce both of those behaviors on my demo.

GuillaumeGomez commented:
> > Also: can you "restore" the previous coloring for the focused search result please?
> 
> Right now the focused search result gets colorful backgrounds, and the mouse hover gets gray. I'd like focus vs hover to be visually distinct. I could switch those, so hover is colorful and focus is gray. Or I could make both colorful, but hover desaturated (but I think this will produce visual confusion).

For me there is only the normal focus (border with pointed line). I'd prefer something a bit more colorful, at least enough to see it and not just the pointed border.

> > scroll with arrows isn't working anymore, we have to use the mouse now
> 
> I'm confused in the same way @Manishearth was. For me, both search result selection and page scrolling with the arrow keys works the same as on master (using FF and Chrome on Linux). Can you tell me the keystrokes you're doing that cause a page scroll on master, but don't cause a page scroll in this branch? What I see is this: search input or search results focused -> down arrow selects a result. No focus -> down arrow scrolls the page.

No sorry, you're both right. Just double checked and when the focus isn't on the search input or on one of the result, it's working as expected.

> > Being able to switch tabs when the search input is focused (using the current shortcut CTRL+up/down) would be nice, because for now it'd force to have a search result focused, which isn't super intuitive.
> > It's very common for me to switch between two tabs and it'd be annoying to have to go through all the items again.
> 
> I don't fully understand your use case here. Can you tell me more? What's an example search where you would be likely to focus an element in one tab, go to another tab, come back to the original tab and want to continue where you left off? Are these very long lists of search results?

It happens when I'm looking for a type to be both in parameters and in the returned value. It's mostly on crates composed of a lot of similar functions with hard to read signatures (someone said simd?). So not sure how common my case is.

bors commented:
:umbrella: The latest upstream changes (presumably #84490) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

jsha commented:
> For me there is only the normal focus (border with pointed line). I'd prefer something a bit more colorful, at least enough to see it and not just the pointed border.

I can't reproduce this, on FF, Chrome, or Edge. Here's what I see:

![image](https://user-images.githubusercontent.com/220205/116004890-a4f53c00-a5b9-11eb-8407-e57c1d3c5a57.png)

In what browser are you seeing the focus without color? And what mouse clicks / keystrokes do you use to get there?

jsha commented:
> Another important point: when we go back to the search input, can you scroll to the top too and not simply focus on the input? We add some issues about this behavior because some browsers only care if the element is in the viewport and not if it's reachable/visible (because hidden under other elements). That could be an issue on docs.rs for example.

The natural way to do this is with `element.scrollIntoView()`. I've tried this locally, and it has a downside: if you're already at the top of the page, it actually scrolls the page down by about a half-line so the search box is at the very tippy-top of the page.

Focusing the search box will generally do the right thing. I can see how there's an issue where a topbar obscures part of the page (like on docs.rs), but I'd rather address that in docs.rs by making the topbar doesn't overlap, but instead shrinks the scrollable part of the page. Overlapping topbars already cause problems: they make page-up / page-down / spacebar scroll by an incorrect amount.

Nemo157 commented:
In `ayu` theme the focus was a simple dotted blue outline, now it appears that both `dark` and `ayu` don't highlight the focused element at all.

jsha commented:
I've pushed some changes and updated the demo. I went ahead and implemented "remember the focused item on the previous tab." It wasn't too complicated and I don't think it will have many unanticipated side effects.

I fixed a nasty bug where I accidentally overrode alt-left and alt-right (forward and back in browser history 😳). And another interesting bug where searching multiple times, or changing crates, would cause jumping by more than one item at a time.

All three themes should now work and show colorful highlighting.

> We can make it so that typing any letter key while the search results are highlight focuses the search input and inserts that letter.

I haven't implemented this yet. Thinking about it some more, I think it will be quite hard to get right. For JS keyboard events we just get a keycode, not an indication of whether something is a letter or punctuation. And the set of keycodes that correspond to letters and punctuation will probably change by locale. My preference right now would be that, when you are navigating through search results, you have to hit 's' to re-focus the search input. Is that okay by you @GuillaumeGomez ?

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2433278128) [(plain)](https://github.com/rust-lang/rust/commit/9798ed32fbe09623c65b8af99462e0282149e6dd/checks/2433278128/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.................................................................................................... 9400/11795
.................................................................................................... 9500/11795
.................................................................................................... 9600/11795
.....................i......i....................................................................... 9700/11795
...................................................................iiiiiii..iiiiii.i................ 9800/11795
.................................................................................................... 10000/11795
.................................................................................................... 10100/11795
.................................................................................................... 10200/11795
.................................................................................................... 10300/11795
---
Suite("src/test/assembly") not skipped for "bootstrap::test::Assembly" -- not in ["src/tools/tidy"]
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 35 tests
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiii....

 finished in 0.158 seconds
Suite("src/test/incremental") not skipped for "bootstrap::test::Incremental" -- not in ["src/tools/tidy"]
Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
---
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 116 tests
iiiiiiiiii.i.i..i...i.ii....i.i....ii..........iiii.........i.....i...i.......ii.i.ii.....iiii.....i 100/116
test result: ok. 78 passed; 0 failed; 38 ignored; 0 measured; 0 filtered out; finished in 2.30s

 finished in 2.367 seconds
Suite("src/test/ui-fulldeps") not skipped for "bootstrap::test::UiFullDeps" -- not in ["src/tools/tidy"]
---
    Finished release [optimized] target(s) in 0.54s
rustdoc: [check-theme] Starting tests! (Ignoring all other arguments)
 - Checking "/checkout/src/librustdoc/html/static/themes/dark.css"... OK
 - Checking "/checkout/src/librustdoc/html/static/themes/ayu.css"... FAILED
  Missing ".search-input:focus" rule

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rustdoc-themes" "/checkout/obj/build/bootstrap/debug/rustdoc" "/checkout/src/librustdoc/html/static/themes"
expected success, got: exit code: 1

```

</details>

GuillaumeGomez commented:
> I haven't implemented this yet. Thinking about it some more, I think it will be quite hard to get right. For JS keyboard events we just get a keycode, not an indication of whether something is a letter or punctuation. And the set of keycodes that correspond to letters and punctuation will probably change by locale. My preference right now would be that, when you are navigating through search results, you have to hit 's' to re-focus the search input. Is that okay by you @GuillaumeGomez ?

I am fine with it.

When I hover with the mouse, I get:

![Screenshot from 2021-04-26 13-50-15](https://user-images.githubusercontent.com/3050060/116078071-7731f500-a696-11eb-8ee9-1f9ba3b2fdbb.png)

This is not great... Also: why did you update the "selected" entry color style? Wouldn't it be better to keep it as is? And finally: for the mouse hover, wouldn't it be simpler to use the same color as focus but with a lighter/darker tone instead of a completely different one?

jsha commented:
> Also: why did you update the "selected" entry color style?

Do you mean the CSS selectors, like `a.result-trait:focus`? Or are you talking about different colors? I think the colors for focused items (previously "highlighted" items) are the same.

> When I hover with the mouse, I get:

Ah, looks like I need a better hover color for dark themes, thanks.

> And finally: for the mouse hover, wouldn't it be simpler to use the same color as focus but with a lighter/darker tone instead of a completely different one?

I think that would require a whole addition set of styles for the hover colors, which would add some complexity and maintenance burden. But more importantly, I think having two colorful selections, even with a lighter/darker tone, would make it confusing as to which is hovered and which is focused.

GuillaumeGomez commented:
> > Also: why did you update the "selected" entry color style?
> 
> Do you mean the CSS selectors, like `a.result-trait:focus`? Or are you talking about different colors? I think the colors for focused items (previously "highlighted" items) are the same.

Nevermind, just realized that I had some "personal" style over the default one. Removed it and back to normal, my bad.

> > And finally: for the mouse hover, wouldn't it be simpler to use the same color as focus but with a lighter/darker tone instead of a completely different one?
> 
> I think that would require a whole addition set of styles for the hover colors, which would add some complexity and maintenance burden. But more importantly, I think having two colorful selections, even with a lighter/darker tone, would make it confusing as to which is hovered and which is focused.

Then why not using the same color and maybe add a "border effect" for the focused one (or not and simply use both same colors both the focus and hover)?

GuillaumeGomez commented:
Before I forget: you'll also need to update the help popup. :)

dns2utf8 commented:
Hi, very cool changes so far

I noticed that the "no results" information on the bottom is visible twice when I search for "fs":
![image](https://user-images.githubusercontent.com/739070/116127810-c5a3bb80-a6b7-11eb-89d9-f8a490c0424e.png)

And three times if the search turns up nothing:
![image](https://user-images.githubusercontent.com/739070/116127998-fd126800-a6b7-11eb-9120-895e217d4fde.png)

dns2utf8 commented:
Another small thing I noticed: when scrolling down with the arrows on firefox and chromium the target URL blocks the view to the currently selected entry.
When using the mouse, the browser moves the info box out of the way, so I am not sure if there is much to be done except maybe move the last entry one more line when at the bottom of a page?

![image](https://user-images.githubusercontent.com/739070/116129292-778fb780-a6b9-11eb-9c56-2e6867e28101.png)

dns2utf8 commented:
I implemented the scrolling ahead functionality [on top of your PR](https://github.com/jsha/rust/pull/1).

It scrolls ahead and prevents that the user can not see the current selection when the browser shows the target URL:
![image](https://user-images.githubusercontent.com/739070/116416506-ed1a9580-a829-11eb-9964-41d391201453.png)

dns2utf8 commented:
Another thing I noticed:
If an entry is selected and the browser window looses focus the keyboard selection gets hidden until the window regains the focus again.
In the current stable the selection stays visible. Can we keep this behaviour or something similar?

jsha commented:
@dns2utf8 Thanks for the PR! In reviewing it, I searched MDN for some scroll-related properties, and I found `scroll-margin`, which seems to do just what we want, so I pushed a commit that does that: https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin. I've also updated the demo.

> If an entry is selected and the browser window looses focus the keyboard selection gets hidden until the window regains the focus again.
> In the current stable the selection stays visible. Can we keep this behaviour or something similar?

I think this would be somewhat messy. From the browser's point of view, the item is not focused if the browser is not focused, so the CSS properties for `:focus` no longer apply. We could probably hack around this by adding an `onfocus` handler that adds the `highlighted` class to search result, and an `onblur` that removes it, _but only if nothing else on the page is becoming focused. However, this seems to eliminate some of the simplicity gains we gt from using the browser's focus mechanism instead of emulating it ourselves.

GuillaumeGomez commented:
@jsha The support for `scroll-margin` seems incomplete on safari. Have you tested on a mac? If not I can, please update your demo link so I can confirm. :)

jsha commented:
Oops, I thought I'd updated the demo. Done now.

BTW [MDN says, under Safari support for scroll-margin](https://developer.mozilla.org/en-US/docs/Web/CSS/scroll-margin): 

> Scroll margin is not applied for scrolls to fragment target or scrollIntoView(), see bug 189265.

I think neither of these apply here.

GuillaumeGomez commented:
I can confirm that it's working perfectly on mac. 👍

GuillaumeGomez commented on src/librustdoc/html/static/search.js:
```diff
@@ -879,106 +879,31 @@ window.initSearch = function(rawSearchIndex) {
-                i_e.onmouseover = mouseover_func;
-            });
-        });
+    function tabLeft() {
```
Instead of having two functions to handle that, wouldn't it be simpler to give `1` or `-1` as parameter to function which would then add it to `currentTab` and then check if `< 0` or `> 2`?

Because the two functions are almost identical.

GuillaumeGomez commented on src/librustdoc/html/static/themes/ayu.css:
```diff
@@ -467,15 +460,38 @@ pre.rust .attribute .ident {}
 pre.rust .kw-2,pre.rust .prelude-ty {}
 .content span.trait,.content a.trait,.block a.current.trait {}
 
+.search-results a:focus {}
```
Maybe use comma instead of `{}` on every line and put the `{}` at the end?

GuillaumeGomez replied:
Ah nevermind, that would error. Is it normal that in ayu they're all the same?

jsha replied:
Yep, see the comment above (prior to FIXME; it's hidden by GitHub):

> /* rules that this theme does not need to set, here to satisfy the rule checker */

GuillaumeGomez commented on src/librustdoc/html/static/themes/ayu.css:
```diff
@@ -467,15 +460,38 @@ pre.rust .attribute .ident {}
+a.result-primitive:focus {}
+a.result-keyword:focus {}
+
+
```
Maybe remove some empty lines here? :)

GuillaumeGomez commented:
Still not correct hover:

<img width="1010" alt="Screenshot 2021-04-28 at 20 08 55" src="https://user-images.githubusercontent.com/3050060/116452146-ab0b4700-a85d-11eb-8773-07981a92be0f.png">

GuillaumeGomez commented:
I noticed a change in behaviour: when you are on the first tab and press left, you should be put on the last one, but it's doing nothing. Same when you're on the last tab and press right.

But otherwise, I really love this change. It's so much better than the current handling. \o/

dns2utf8 commented:
I usually prefere a CSS over a JS solution, but the scrolling behaves very strange on my machine. 
On Firefox 88 (Arch Linux) the scroll margin lets the selection out, but it works on chromium -> is it a firefox bug or does one of my addons interfere? 

https://user-images.githubusercontent.com/739070/116457474-f9ffaf00-a852-11eb-83a0-9c8f53fa2832.mp4

GuillaumeGomez commented:
Ah right, in my case on firefox, the focus is completely broken. It goes beyond the last one visible before we can see again.

jsha commented:
Yeah, agreed the current solution is broken. I need to experiment a bit more with scroll-margin (negative vs positive values?) and figure out if it's the right tool for the job. It's possible the proposed JS solution is better.

dns2utf8 commented:
I fetched the latest firefox beta 89.0b5 and the behaviour is the same as with 88 :/

dns2utf8 commented:
I did open an [issue on firefoxs bugzilla #1708303](https://bugzilla.mozilla.org/show_bug.cgi?id=1708303), if you have an account please vote for the issue.

dns2utf8 commented:
A fix for the scroll-margin bug laded in the [Firefox Source](https://hg.mozilla.org/mozilla-central/rev/8befa6dc66fd) today.
I just tested [my minimal example](https://data.estada.ch/firefox_scroll-margin_bug/) with the latest desktop nightly build and it works. So, then in ~12 weeks the fix will be in in Firefox 90 stable.
The android nightly is still broken, I assume the fix will land there tomorrow.

GuillaumeGomez commented:
This is way too recent to be used if it's broken in older versions. We'll need to find another way. :)

dns2utf8 commented:
What if we the JavaScript implementation and the CSS together. Adding a mark to both of them why the feature is implemented twice.
Also, I just checked [caniuse](https://caniuse.com/?search=scroll-margin) and it said that IE does not support it at all.

Currently, the JS implementation emulates `scroll-margin-bottom: 1em;` and the CSS is was at `300px`. That way browsers with (correct) support will present a smoother experience and works on scroll up as well.
Or just make a note for 2022?

GuillaumeGomez commented:
We dropped IE support when we decided to use `<details>` and `<summary>`, so not a problem on that. :)

Maybe we just need to call `elem.focus()` and that would do the trick?

jsha commented:
> What if we the JavaScript implementation and the CSS together.

I think the JS implementation is fine, and I don't want to take the extra complexity for two implementations of the same thing.

> Currently, the JS implementation emulates scroll-margin-bottom: 1em; and the CSS is was at 300px.

If you prefer a larger margin, I can make the JS do that. The current version was a fairly arbitrary choice.

@GuillaumeGomez, following up on your previous request to make both the hover and the focus colorful:

 > > > And finally: for the mouse hover, wouldn't it be simpler to use the same color as focus but with a lighter/darker tone instead of a completely different one?

> > I think that would require a whole addition set of styles for the hover colors, which would add some complexity and maintenance burden. But more importantly, I think having two colorful selections, even with a lighter/darker tone, would make it confusing as to which is hovered and which is focused.

> Then why not using the same color and maybe add a "border effect" for the focused one (or not and simply use both same colors both the focus and hover)?

I really think the bright colors for focus and grey for hover are the right combination. I'm partly taking an example from Firefox here:

<img width=300 src=https://user-images.githubusercontent.com/220205/117585412-30e18a80-b0c7-11eb-9cdf-aec7a8060d65.png>

But also my reasoning is: There's a good chance that your mouse will "accidentally" hover something because it's just sitting around; it should have a less distracting color than the keyboard focus for that reason.

dns2utf8 commented:
Re the scrolling, I am happy with it. Sometimes the distance to the end of the window fluctuates by a few pixels but it kepps the math simple so I would keep it simple for now :)

jsha commented:
> when an item is selected, its text becomes white

Fixed!

GuillaumeGomez commented:
Just realized something funny: the further down we go, the closer to the bottom of the window the focus is. Can be fixed later as it's not really a big issue, this PR has been open for long enough and I'd really like to see it merged. :D

@dns2utf8 One last check from you too please? :)

bors commented:
:umbrella: The latest upstream changes (presumably #85258) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

GuillaumeGomez commented:
Let's go forward with this then. Really can't wait for this improvement to get merged! :)

@bors: r+

bors commented:
:pushpin: Commit b615c0c85469c94041a5e68b9d8b68dcf799f9f1 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez b615c0c85469c94041a5e68b9d8b68dcf799f9f1 -->
<!-- homu: {"type":"Approved","sha":"b615c0c85469c94041a5e68b9d8b68dcf799f9f1","approver":"GuillaumeGomez"} -->
